### ISSUE TITLE
[PATCH] Regression in 5.12: labelled string eval

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10301]
Migrated from [rt.perl.org#74290](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=74290) (status was 'resolved')

Searchable as RT74290$

**Comment by p5pRT** [comment]
### From @cpansprout

http&#8203;://www\.nntp\.perl\.org/group/perl\.perl5\.porters/2010/04/msg158760\.html

This was broken by change 33656 \(http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff/dca6062a863d0 
\) moves labels to the hints hash\.

Here’s another example of the same bug\, which shows that that it has  
nothing to do with $\_&#8203;:

perl5\.12\.0 \-le 'F1&#8203;:\+\+$x and eval q"print qq/x is $x/;  exit if \+\+$y ==  
10; goto q/F1/;"'
x is 1
x is 1
x is 1
x is 1
x is 1
x is 1
x is 1
x is 1
x is 1
x is 1

In short\,
&nbsp;    Foo&#8203;: eval "goto Foo"
doesn’t go to Foo\, but simply repeats the eval\.

I tried making S\_dofindlabel skip any op that == PL\_eval\_start\, but  
that prevents eval "Foo&#8203;: print q/foo/; goto Foo" from working\.

In other words\, the goto inside the eval cannot tell the difference  
between
&nbsp;    Foo&#8203;: eval " \.\.\. "
and
&nbsp;    eval "Foo&#8203;: \.\.\. "

It seems to have something to do with %^H not being synchronised with  
the equivalent internal structure\, because putting BEGIN\{ delete $^H 
\{anything\} \} at the beginning of the evalled code fixes it\.

The attached patch is perhaps not the best fix\, but works around the  
bug in a way that can’t cause any negative side\-effects\.

Use of uninitialized value $category in concatenation \(\.\) or string  
at /usr/local/bin/perlbug5\.12\.0 line 645\.
Use of uninitialized value $severity in concatenation \(\.\) or string  
at /usr/local/bin/perlbug5\.12\.0 line 645\.


---
Flags&#8203;:
&nbsp;     category=core
&nbsp;     severity=medium


---
Site configuration information for perl 5\.12\.0&#8203;:

Configured by sprout at Tue Apr  6 15&#8203;:41&#8203;:56 PDT 2010\.

Summary of my perl5 \(revision 5 version 12 subversion 0 patch v5\.12\.0\- 
RC3\-6\-g8c57606\) configuration&#8203;:
&nbsp;   Snapshot of&#8203;: 431be7b35e7448c1ced59444d6bdcfa11c5f83c7
&nbsp;   Platform&#8203;:
&nbsp;     osname=darwin\, osvers=10\.0\.0\, archname=darwin\-2level
&nbsp;     uname='darwin pint\.local 10\.0\.0 darwin kernel version 10\.0\.0&#8203;: fri  
jul 31 22&#8203;:47&#8203;:34 pdt 2009; root&#8203;:xnu\-1456\.1\.25~1release\_i386 i386 '
&nbsp;     config\_args='\-de \-Dusedevel'
&nbsp;     hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;     useithreads=undef\, usemultiplicity=undef
&nbsp;     useperlio=define\, d\_sfio=undef\, uselargefiles=define\,  
usesocks=undef
&nbsp;     use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
&nbsp;     usemymalloc=n\, bincompat5005=undef
&nbsp;   Compiler&#8203;:
&nbsp;     cc='cc'\, ccflags ='\-fno\-common \-DPERL\_DARWIN \-no\-cpp\-precomp \-fno\- 
strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'\,
&nbsp;     optimize='\-O3'\,
&nbsp;     cppflags='\-no\-cpp\-precomp \-fno\-common \-DPERL\_DARWIN \-no\-cpp\- 
precomp \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/ 
include'
&nbsp;     ccversion=''\, gccversion='4\.2\.1 \(Apple Inc\. build 5646\)'\,  
gccosandvers=''
&nbsp;     intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;     ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\,  
Off\_t='off\_t'\, lseeksize=8
&nbsp;     alignbytes=8\, prototype=define
&nbsp;   Linker and Libraries&#8203;:
&nbsp;     ld='env MACOSX\_DEPLOYMENT\_TARGET=10\.3 cc'\, ldflags =' \-fstack\- 
protector \-L/usr/local/lib'
&nbsp;     libpth=/usr/local/lib /usr/lib
&nbsp;     libs=\-ldbm \-ldl \-lm \-lutil \-lc
&nbsp;     perllibs=\-ldl \-lm \-lutil \-lc
&nbsp;     libc=/usr/lib/libc\.dylib\, so=dylib\, useshrplib=false\,  
libperl=libperl\.a
&nbsp;     gnulibc\_version=''
&nbsp;   Dynamic Linking&#8203;:
&nbsp;     dlsrc=dl\_dlopen\.xs\, dlext=bundle\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;     cccdlflags=' '\, lddlflags=' \-bundle \-undefined dynamic\_lookup \-L/ 
usr/local/lib \-fstack\-protector'

Locally applied patches&#8203;:
&nbsp;     RC5



---
@&#8203;INC for perl 5\.12\.0&#8203;:
&nbsp;     /usr/local/lib/perl5/site\_perl/5\.12\.0/darwin\-2level
&nbsp;     /usr/local/lib/perl5/site\_perl/5\.12\.0
&nbsp;     /usr/local/lib/perl5/5\.12\.0/darwin\-2level
&nbsp;     /usr/local/lib/perl5/5\.12\.0
&nbsp;     /usr/local/lib/perl5/site\_perl
&nbsp;     \.



---
Environment for perl 5\.12\.0&#8203;:
&nbsp;     DYLD\_LIBRARY\_PATH \(unset\)
&nbsp;     HOME=/Users/sprout
&nbsp;     LANG=en\_US\.UTF\-8
&nbsp;     LANGUAGE \(unset\)
&nbsp;     LD\_LIBRARY\_PATH \(unset\)
&nbsp;     LOGDIR \(unset\)
&nbsp;     PATH=/usr/bin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/usr/local/bin&#8203;:/usr/X11/bin&#8203;:/ 
usr/local/bin
&nbsp;     PERL\_BADLANG \(unset\)
&nbsp;     SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @cpansprout

<details><summary>Inline Patch</summary>

```diff
diff -Nurp blead/pp_ctl.c blead-labelled-eval/pp_ctl.c
--- blead/pp_ctl.c	2010-04-10 14:03:15.000000000 -0700
+++ blead-labelled-eval/pp_ctl.c	2010-04-11 14:27:16.000000000 -0700
@@ -3763,6 +3763,8 @@ PP(pp_entereval)
 	HINTS_REFCNT_LOCK;
 	PL_compiling.cop_hints_hash->refcounted_he_refcnt++;
 	HINTS_REFCNT_UNLOCK;
+	PL_compiling.cop_hints_hash
+	    = Perl_store_cop_label(aTHX_ PL_compiling.cop_hints_hash, "");
     }
     /* special case: an eval '' executed within the DB package gets lexically
      * placed in the first non-DB CV rather than the current CV - this
diff -Nurp blead/t/op/eval.t blead-labelled-eval/t/op/eval.t
--- blead/t/op/eval.t	2010-04-02 12:23:09.000000000 -0700
+++ blead-labelled-eval/t/op/eval.t	2010-04-11 14:30:00.000000000 -0700
@@ -6,7 +6,7 @@ BEGIN {
     require './test.pl';
 }
 
-print "1..106\n";
+print "1..107\n";
 
 eval 'print "ok 1\n";';
 
@@ -602,3 +602,18 @@ eval q{ eval { + } };
 print "ok\n";
 EOP
 
+fresh_perl_is(<<'EOP', <<'expected', undef, 'labelled string eval');
+#!perl -l
+F1:++$x and eval q"print qq/x is $x/; exit if ++$y == 10; goto q/F1/;"
+EOP
+x is 1
+x is 2
+x is 3
+x is 4
+x is 5
+x is 6
+x is 7
+x is 8
+x is 9
+x is 10
+expected
```
</details>

**Comment by p5pRT** [comment]
### From @nwc10

On Sun\, Apr 11\, 2010 at 05&#8203;:03&#8203;:41PM \-0700\, Father Chrysostomos wrote&#8203;:


> diff \-Nurp blead/pp\_ctl\.c blead\-labelled\-eval/pp\_ctl\.c
> \-\-\- blead/pp\_ctl\.c    2010\-04\-10 14&#8203;:03&#8203;:15\.000000000 \-0700
> \+\+\+ blead\-labelled\-eval/pp\_ctl\.c    2010\-04\-11 14&#8203;:27&#8203;:16\.000000000 \-0700
> @&#8203;@&#8203; \-3763\,6 \+3763\,8 @&#8203;@&#8203; PP\(pp\_entereval\)
>      HINTS\_REFCNT\_LOCK;
>      PL\_compiling\.cop\_hints\_hash\->refcounted\_he\_refcnt\+\+;
>      HINTS\_REFCNT\_UNLOCK;
> \+    PL\_compiling\.cop\_hints\_hash
> \+        = Perl\_store\_cop\_label\(aTHX\_ PL\_compiling\.cop\_hints\_hash\, ""\);
>      \}
>      /\* special case&#8203;: an eval '' executed within the DB package gets lexically
>       \* placed in the first non\-DB CV rather than the current CV \- this

As is\, that's going to leak\, as the reference to the \(then current\)
PL\_compiling\.cop\_hints\_hash is going to be bumped once by
PL\_compiling\.cop\_hints\_hash\->refcounted\_he\_refcnt\+\+ and then bumped again
within Perl\_store\_cop\_label\(\)\.

The simplest solution would be to remove the preceding 3 lines\.

A more efficient solution \(avoiding any allocation\) would be to walk back
to the parent if the topmost item on PL\_compiling\.cop\_hints\_hash is a "label"\.

Nicholas Clark

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @nwc10

On Tue\, Apr 13\, 2010 at 11&#8203;:31&#8203;:51AM \+0100\, Nicholas Clark wrote&#8203;:
> On Sun\, Apr 11\, 2010 at 05&#8203;:03&#8203;:41PM \-0700\, Father Chrysostomos wrote&#8203;:
> 
> 
> > diff \-Nurp blead/pp\_ctl\.c blead\-labelled\-eval/pp\_ctl\.c
> > \-\-\- blead/pp\_ctl\.c    2010\-04\-10 14&#8203;:03&#8203;:15\.000000000 \-0700
> > \+\+\+ blead\-labelled\-eval/pp\_ctl\.c    2010\-04\-11 14&#8203;:27&#8203;:16\.000000000 \-0700
> > @&#8203;@&#8203; \-3763\,6 \+3763\,8 @&#8203;@&#8203; PP\(pp\_entereval\)
> >      HINTS\_REFCNT\_LOCK;
> >      PL\_compiling\.cop\_hints\_hash\->refcounted\_he\_refcnt\+\+;
> >      HINTS\_REFCNT\_UNLOCK;
> > \+    PL\_compiling\.cop\_hints\_hash
> > \+        = Perl\_store\_cop\_label\(aTHX\_ PL\_compiling\.cop\_hints\_hash\, ""\);
> >      \}
> >      /\* special case&#8203;: an eval '' executed within the DB package gets lexically
> >       \* placed in the first non\-DB CV rather than the current CV \- this
> 
> As is\, that's going to leak\, as the reference to the \(then current\)
> PL\_compiling\.cop\_hints\_hash is going to be bumped once by
> PL\_compiling\.cop\_hints\_hash\->refcounted\_he\_refcnt\+\+ and then bumped again
> within Perl\_store\_cop\_label\(\)\.
> 
> The simplest solution would be to remove the preceding 3 lines\.

I'm wrong\. Your patch as is is correct\.

Thanks for correctly identifying a solution\, and for the simpler test code\.

> A more efficient solution \(avoiding any allocation\) would be to walk back
> to the parent if the topmost item on PL\_compiling\.cop\_hints\_hash is a "label"\.

However\, I preferred this approach\, and implemented it as 47550813adf9ff40\.

Nicholas Clark

**Comment by p5pRT** [comment]
@nwc10 - Status changed from 'open' to 'resolved'
